src/api/dirmonitor/posix_polling.c

```diff
@@ -0,0 +1,142 @@
+#include <stdlib.h>
+#include <string.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <time.h>
+#include <SDL.h>
+
+#define EVENT_MODIFIED (1 << 0)
+
+#if _BSD_SOURCE || _SVID_SOURCE || _XOPEN_SOURCE > 700 || _POSIX_C_SOURCE >= 200809L
+#define STAT_HAS_TIM
+#endif
+
+struct file_info {
+  int fd;
+  dev_t dev;
+  ino_t ino;
+  struct timespec mtim;
+  int events;
+};
+
+struct dirmonitor_internal {
+  struct file_info* fds;
+  size_t fd_count;
+};
+
+  
+struct dirmonitor_internal* init_dirmonitor() {
+  struct dirmonitor_internal* monitor = calloc(1, sizeof(struct dirmonitor_internal));
+
+  monitor->fd_count = 0;
+  monitor->fds = NULL;
+  return monitor;
+}
+
+void deinit_dirmonitor(struct dirmonitor_internal* monitor) {
+  for (size_t i = 0; i < monitor->fd_count; ++i)
+    close(monitor->fds[i].fd);
+  free(monitor->fds);
+}
+
+int get_changes_dirmonitor(struct dirmonitor_internal* monitor, char* buffer, int len)
+{
+  struct stat new_stat;
+  int c = 0;
+
+  for (size_t i = 0; i < monitor->fd_count; ++i) {
+    if (monitor->fds[i].fd < 0)
+      continue;
```

Comment:
Could `fd_count` be raced over?
Same with the `fds` structs.

For example:
1. We enter `get_changes_dirmonitor`
2. Main thread enters `add_dirmonitor`
3. `add_dirmonitor` needs to allocate new space, so it does, but still doesn't switch it out
4. `get_changes_dirmonitor` saves `fds` in a register
5. `add_dirmonitor` switches out the spaces, frees the old `fds`
6. `get_changes_dirmonitor` uses the `fds` from the register, SIGSEGVs.